nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/QuerySalesforceObject.java

```diff
@@ -387,11 +461,63 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
+        String customQuery = context.getProperty(CUSTOM_SOQL_QUERY).evaluateAttributeExpressions(flowFile).getValue();
+        do {
+            FlowFile outgoingFlowFile;
+            if (flowFile != null) {
+                outgoingFlowFile = session.create(flowFile);
+            } else {
+                outgoingFlowFile = session.create();
+            }
+            Map<String, String> attributes = new HashMap<>();
+            try (InputStream response = getResultInputStream(nextRecordsUrl.get(), customQuery)) {
+                outgoingFlowFile = session.write(outgoingFlowFile, parseHttpResponse(response, nextRecordsUrl, totalSize));
+                int recordCount = nextRecordsUrl.get() != null ? 2000 : Integer.parseInt(totalSize.get()) % 2000;
+                attributes.put(CoreAttributes.MIME_TYPE.key(), "application/json");
+                attributes.put(TOTAL_RECORD_COUNT, String.valueOf(recordCount));
+                session.adjustCounter("Salesforce records processed", recordCount, false);
+                session.putAllAttributes(outgoingFlowFile, attributes);
+                session.transfer(outgoingFlowFile, REL_SUCCESS);
+            } catch (IOException e) {
+                session.remove(outgoingFlowFile);
+                throw new ProcessException("Couldn't get Salesforce records", e);
+            }
+        } while (nextRecordsUrl.get() != null);
+        if (flowFile != null) {
+            session.remove(flowFile);
+        }
+    }
+
+    private OutputStreamCallback parseHttpResponse(final InputStream in, AtomicReference<String> nextRecordsUrl, AtomicReference<String> totalSize) {
+        nextRecordsUrl.set(null);
+        return out -> {
+            try (JsonParser jsonParser = JSON_FACTORY.createParser(in);
+                 JsonGenerator jsonGenerator = JSON_FACTORY.createGenerator(out, JsonEncoding.UTF8)) {
+                while (jsonParser.nextToken() != null) {
+                    if (jsonParser.getCurrentToken() == JsonToken.FIELD_NAME && jsonParser.getCurrentName()
+                            .equals(TOTAL_SIZE)) {
+                        jsonParser.nextToken();
+                        totalSize.set(jsonParser.getValueAsString());
+                    } else if (jsonParser.getCurrentToken() == JsonToken.FIELD_NAME && jsonParser.getCurrentName()
+                            .equals(NEXT_RECORDS_URL)) {
+                        jsonParser.nextToken();
+                        nextRecordsUrl.set(jsonParser.getValueAsString());
+                    } else if (jsonParser.getCurrentToken() == JsonToken.FIELD_NAME && jsonParser.getCurrentName()
+                            .equals(RECORDS)) {
+                        jsonParser.nextToken();
+                        jsonGenerator.copyCurrentStructure(jsonParser);
+                    }
```

Comment:
Bit of a code duplication here.
I would consider something like
```suggestion
                while (jsonParser.nextToken() != null) {
                    if (nextTokenIs(jsonParser, JsonToken.FIELD_NAME, TOTAL_SIZE)) {
                        totalSize.set(jsonParser.getValueAsString());
                    } else if (nextTokenIs(jsonParser, JsonToken.FIELD_NAME, NEXT_RECORDS_URL)) {
                        nextRecordsUrl.set(jsonParser.getValueAsString());
                    } else if (nextTokenIs(jsonParser, JsonToken.FIELD_NAME, RECORDS)) {
                        jsonGenerator.copyCurrentStructure(jsonParser);
                    }
```
with
```java
    private boolean nextTokenIs(JsonParser jsonParser, JsonToken fieldName, String value) throws IOException {
        return jsonParser.getCurrentToken() == fieldName && jsonParser.getCurrentName()
                .equals(value) && jsonParser.nextToken() != null;
    }
```
I do something uncommon here because this boolean method not only checks something but also moves to the next token if the corresponding check evaluates to true. But this type of behaviour in parsers are usually acceptable.

Reply:
This is a good idea. I'd leave the fieldName parameter out for now as it's value is always JsonToken.FIELD_NAME. Thanks!